### PR TITLE
[vla, train] feat: add error feedback and selective-precision exemption to fp8 DDP comm hook

### PR DESCRIPTION
**Pull request by XueSongTap**
## Summary

Two independent, optional precision recoveries for `fp8_a2a_allgather_hook`, each off by default so existing runs are bit-for-bit unchanged:

1. **Selective-precision exemption** (`--ddp-comm-hook-fp8-exempt`, recommended) — keep loss-sensitive parameters at the exact fp32 cross-rank mean instead of fp8. fp8 block quantization compresses 1-D RMSNorm/LayerNorm scales and biases ~10-20x harder than 2-D weights (they share 256-element blocks with larger-magnitude neighbours), so exempting ~0.29% of elements removes the hook's systematic per-step bias. Over 300 steps of GR00T-N1.7 this cuts the full-run mean |rel-error| from 10.87% to 2.13% at plain-hook throughput and memory (no extra scratch).
2. **Error feedback** (`--ddp-comm-hook-fp8-error-feedback`) — carry each quantization point's residual into the next step so the accumulated bias telescopes to `e_0 - e_T` instead of drifting. Cuts the 300-step tail loss bias from +4.19% to within the ±0.4pp run-to-run noise floor, at ~3% throughput and +6.8 GiB scratch.

Both recover the loss to a comparable level; exemption does so at lower throughput and memory cost, so it is the recommended default. They are orthogonal and can be combined.

### Selective-precision exemption
- `fp8_a2a_comm.py`:
  - `--ddp-comm-hook-fp8-exempt`, a comma list of `{1d,embed,head}` (default `1d,embed`, empty = off), validated against the token whitelist in `configure()` and stored in module state alongside the other `ddp_comm_hook_fp8_*` tunables.
  - Matching is **structural, not name-substring**, so it is model-agnostic: `1d` = every 1-D tensor (norm scales, biases); `embed` = params owned by an `nn.Embedding` module; `head` = params owned by an output-head leaf (`lm_head`/`output`/`score`/`classifier`). This avoids two substring pitfalls — a top-level container merely named `action_head` pulling in all of its parameters, and a `timestep_embedder` MLP being treated as an embedding.
  - `set_param_names()` walks `named_modules()` at DDP-wrap time (the comm hook only sees `(state, bucket)`) and records `id(param) -> name` plus the embedding/head parameter id sets; `_exempt_indices`/`_exempt_buffers` cache the per-bucket exempt positions and reduction buffers, rebuilt on bucket-layout change like `_scratch_for`.
  - In the hook, snapshot the exempt positions before the fp8 path mutates the buffer, reduce them with a synchronous inline fp32 `all_reduce` (issued in the same deterministic bucket order as the a2a/all-gather), then write the exact mean back after the all-gather. All buffers are pre-allocated, so the correction is CUDA-graph capture-safe. Exemption forces the synchronous inline path (an async continuation let the third collective interleave nondeterministically across ranks).
  - The per-bucket exempt fraction is logged once at INFO, so a large trainable vocab embedding blowing up the exempt set (e.g. `tune_llm=True`) is visible without a hard cap.
- `training_args.py`: add the `ddp_comm_hook_fp8_exempt` arg.
- `parallel.py`: pass it into `fp8_a2a_comm.configure()` and call `fp8_a2a_comm.set_param_names(model)` before registering the hook.

### Error feedback
- `fp8_a2a_comm.py`:
  - `--ddp-comm-hook-fp8-error-feedback` modes `none/allgather/both`, plumbed through `configure()` and per-bucket scratch. Residuals are fp32 (a bf16 residual reintroduces most of the drift EF removes) and counted against the scratch budget; dropped with the scratch when the bucket layout changes.
  - Fused residual kernel `residual = value - dequant(q)` closes each EF step in one pass instead of two full-bucket HBM round trips.
  - Fix a latent quantization-kernel bug the EF path exposes: a block whose `amax/448` lands in fp32 denormals made `1/scale` overflow to `+inf` and every element convert to fp8 NaN, which EF then latched forever. Guard on the fp32 normal-range lower bound and flush such a block to zero. Only reachable with EF on, so precision-only runs are unaffected.
- `training_args.py`: add the `ddp_comm_hook_fp8_error_feedback` arg.
- `parallel.py`: pass it into `fp8_a2a_comm.configure()`.

### CODE REVIEW

**Comment by XueSongTap** [comment]
@claude review this PR

File: loongforge/embodied/distributed/ddp_utils/fp8_a2a_comm.py [hunk 1]
```diff
@@ -424,21 +745,132 @@ def fp8_a2a_allgather_hook(process_group, bucket):
...
+
+        ``tensor`` is safe to modify in place: it is the bucket buffer, and the
+        AllGather result overwrites it wholesale at the end of the hook.
+        """
+        if st.res1 is not None:
+            tensor.add_(st.res1)
+        quantize_chunks(tensor, st.send, st.numel, S, chunk_u8, world_size, block)
+        if st.res1 is not None:
+            ef_residual(st.send, st.res1, tensor, st.numel, S, chunk_u8,
+                        world_size, block)
```
[the remark this answers is not in the document]
Comment (XueSongTap) [reply loongforge/embodied/distributed/ddp_utils/fp8_a2a_comm.py:870]: Fixed in fa10638: zero both EF residuals in patch_exempt after writing the exact mean — res1 at exempt_idx, res2 at the shard-local offsets (new cached _exempt_shard_indices). Verified 300-iter hook-ef-exempt vs baseline: no growing bias.

**Review by nullnonenilNULL — APPROVED** [review]
